### PR TITLE
ceph-dencoder: add support for RGWLifecycleConfiguration

### PR DESCRIPTION
**Pull request by cooboos**
Signed-off-by: Songbo Wang wangsongbo@cloudin.cn

### CODE REVIEW

File: src/rgw/rgw_json_enc.cc [hunk 1]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  encode_json("filter", filter, f);
+  encode_json("dm_expiration", dm_expiration, f);
+}
+
+void RGWLifecycleConfiguration::dump(Formatter *f) const
+{
+  map<string, lc_op>::const_iterator iter = prefix_map.begin();
+  f->open_array_section("prefix_map");
+  for (; iter != prefix_map.end(); ++iter) {
+    f->dump_string("prefix", iter->first);
```
Comment (liewegas) [inline src/rgw/rgw_json_enc.cc]: the items in the array shoudl all have the same type.  if you're trying to make a map, do open_object_section outside the loop, and inside the loop to open_object_section(iter->first.c_str()).

Comment (cooboos) [reply src/rgw/rgw_json_enc.cc]: Thanks. I have correct it. 
But I have got the same mismatch type  in both RGWObjManifest::dump  and RGWAccessControlList::dump;Should I pull another new request to correct that ?

Comment (cooboos) [reply src/rgw/rgw_json_enc.cc]: sorry @liewegas  , It's my mistake. I misunderstood you suggestion,  I have correct it.

File: src/rgw/rgw_json_enc.cc [hunk 2]
```diff
@@ -1494,3 +1494,80 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  multimap<string, LCRule>::const_iterator multi_iter = rule_map.begin();
+  f->open_array_section("rule_map");
+  for (; multi_iter != rule_map.end(); ++multi_iter) {
+    f->open_object_section("entry");
+    f->dump_string("id", multi_iter->first);
+    f->open_object_section("rule");
+    multi_iter->second.dump(f);
+    f->close_section();
+    f->close_section();
+  }
```
Comment (liewegas) [inline src/rgw/rgw_json_enc.cc]: it's a matter of taste, but you can also do
<pre>
+  f->open_object_section("rule_map");
 +  for (; multi_iter != rule_map.end(); ++multi_iter) {
 +    f->open_object_section(multi_iter->first.c_str());
 +    multi_iter->second.dump(f);
 +    f->close_section();
 +  }
 +  f->close_section();
</pre>

Comment (liewegas) [reply src/rgw/rgw_json_enc.cc]: or even
<pre>
+  f->open_object_section("rule_map");
 +  for (; multi_iter != rule_map.end(); ++multi_iter) {
 +    f->dump_object(multi_iter->first.c_str(), multi_iter->second);
 +  }
 +  f->close_section();
</pre>

Comment (cooboos) [reply src/rgw/rgw_json_enc.cc]: @liewegas  Thank you for your suggestion, I have fixed it. Please check,  when you get a chance.

File: src/rgw/rgw_json_enc.cc [hunk 3]
```diff
@@ -1494,3 +1494,70 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+{
+  map<string, lc_op>::const_iterator iter = prefix_map.begin();
+  f->open_object_section("prefix_map");
+  for (; iter != prefix_map.end(); ++iter) {
+    f->dump_object(iter->first.c_str(), iter->second);
+  }
+  f->close_section();
+
+  multimap<string, LCRule>::const_iterator multi_iter = rule_map.begin();
+  f->open_object_section("rule_map");
```
Comment (liewegas) [inline src/rgw/rgw_json_enc.cc]: here using an object isn't a good idea since it's a multimap<>, so there might be duplicate rule names?

Comment (cooboos) [reply src/rgw/rgw_json_enc.cc]: @liewegas Thank you,  Please check again. The last change is the map usging an object , and the multimap using an array.

File: src/rgw/rgw_json_enc.cc [hunk 4]
```diff
@@ -1494,3 +1494,75 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+}
+
+void lc_op::dump(Formatter *f) const
+{
+  encode_json("status", status, f);
+  encode_json("dm_expiration", dm_expiration, f);
+  encode_json("expiration", expiration, f);
+  encode_json("noncur_expiration", noncur_expiration, f);
+  encode_json("mp_expiration", mp_expiration, f);
+  if (expiration_date != boost::none) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: nit, could just put `if (obj_tags)`.

File: src/rgw/rgw_json_enc.cc [hunk 5]
```diff
@@ -1494,3 +1494,75 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+    utime_t ut(*expiration_date);
+    encode_json("expiration_date", ut, f);
+  }
+  if (obj_tags != boost::none) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: ditto.

File: src/rgw/rgw_json_enc.cc [hunk 6]
```diff
@@ -1494,3 +1494,75 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  encode_json("mp_expiration", mp_expiration, f);
+  encode_json("filter", filter, f);
+  encode_json("dm_expiration", dm_expiration, f);
+}
+
+void RGWLifecycleConfiguration::dump(Formatter *f) const
+{
+  map<string, lc_op>::const_iterator iter = prefix_map.begin();
+  f->open_object_section("prefix_map");
+  for (; iter != prefix_map.end(); ++iter) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: use range-based loop, if you please.

File: src/rgw/rgw_json_enc.cc [hunk 7]
```diff
@@ -1494,3 +1494,75 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+    f->dump_object(iter->first.c_str(), iter->second);
+  }
+  f->close_section();
+
+  multimap<string, LCRule>::const_iterator multi_iter = rule_map.begin();
+  f->open_array_section("rule_map");
+  for (; multi_iter != rule_map.end(); ++multi_iter) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: ditto.

File: src/rgw/rgw_lc.h [hunk 8]
```diff
@@ -61,7 +61,7 @@ class LCExpiration
     DECODE_FINISH(bl);
   }
   void dump(Formatter *f) const;
-//  static void generate_test_instances(list<ACLOwner*>& o);
+  //static void generate_test_instances(list<ACLOwner*>& o);
```
Comment (tchaikov) [inline src/rgw/rgw_lc.h]: please drop this change if it's not relevant.

File: src/rgw/rgw_json_enc.cc [hunk 9]
```diff
@@ -1494,3 +1494,75 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+    f->open_object_section("tag_map");
+    f->dump_string("Key", tag.first);
+    f->dump_string("Value", tag.second);
+    f->close_section();
+  }
+}
+
+void lc_op::dump(Formatter *f) const
+{
+  encode_json("status", status, f);
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: why are you hardwiring this to json? can we use formatter's method?

**Comment by cooboos** [comment]
@tchaikov  Thanks for your suggestion, pls check again.

File: src/rgw/rgw_json_enc.cc [hunk 10]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
   encode_json("stage", stage, f);
   f->close_section();
 }
+
+void RGWObjTags::dump(Formatter *f) const
+{
+  for (const auto& tag: tag_map){
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: nit, it's `const` already, b/c this method is `const`.

File: src/rgw/rgw_json_enc.cc [hunk 11]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+{
+  f->dump_bool("status", status);
+  f->dump_bool("dm_expiration", dm_expiration);
+
+  f->dump_int("expiration", expiration);
+  f->dump_int("noncur_expiration", noncur_expiration);
+  f->dump_int("mp_expiration", mp_expiration);
+  if (expiration_date) {
+    utime_t ut(*expiration_date);
+    encode_json("expiration_date", ut, f);
```
[the remark this answers is not in the document]
Comment (cooboos) [reply src/rgw/rgw_json_enc.cc]: @tchaikov  Sorry, but there is no definition about utime_t::dump(). Do I need to define a new utime_t::dump() ? or transform utime_t to string ? or use f->dump_stream ?
appreciate for your comments.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/rgw/rgw_json_enc.cc]: @cooboos ahh, right! i didn't realize that `utime_t` does not have this method ATM. probably a better alternative is `f->dump_stream()`. thanks!

**Comment by cooboos** [comment]
@tchaikov  pls check again.

**Comment by tchaikov** [comment]
also your "Signed-off-by" line does not look right.
```
Signed-off-by: wangsongbo wangsongbo@cloudin.cn
```

see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work

**Comment by cooboos** [comment]
@tchaikov  OK, you  are so appreciated.

**Comment by tchaikov** [comment]
still does not look right, to be specific
```
Signed-off-by: Songbo Wang <wangsongbo@cloudin.cn>
```

please note that the email is quoted using bracket. if your git is configured correctly, `git commit -s` will sign off your commit automatically.

**Comment by cooboos** [comment]
@tchaikov  you are a good reviewer, thank you. Check again pls.

File: src/rgw/rgw_lc.cc [hunk 12]
```diff
@@ -13,6 +13,7 @@
 #include "rgw_common.h"
 #include "rgw_bucket.h"
 #include "rgw_lc.h"
+#include "common/ceph_json.h"
```
Comment (tchaikov) [inline src/rgw/rgw_lc.cc]: why include this header here? the dump() method is defined in  src/rgw/rgw_json_enc.cc.

Comment (cooboos) [reply src/rgw/rgw_lc.cc]: @tchaikov  I defined dump() method in other file previously, then forgot to remove it. so... 
Thanks anyway, check pls.

File: src/rgw/rgw_json_enc.cc [hunk 13]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  f->dump_object("expiration", expiration);
+  f->dump_object("noncur_expiration", noncur_expiration);
+  f->dump_object("mp_expiration", mp_expiration);
+  f->dump_object("filter", filter);
+  f->dump_bool("dm_expiration", dm_expiration);
+}
+
+void RGWLifecycleConfiguration::dump(Formatter *f) const
+{
+  map<string, lc_op>::const_iterator iter = prefix_map.begin();
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: please remove this line.

File: src/rgw/rgw_json_enc.cc [hunk 14]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  f->open_object_section("prefix_map");
+  for (auto& iter : prefix_map) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: `iter` is not an iterator. might want to s/iter/op/.

File: src/rgw/rgw_json_enc.cc [hunk 15]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+    f->dump_object(iter.first.c_str(), iter.second);
+  }
+  f->close_section();
+
+  multimap<string, LCRule>::const_iterator multi_iter = rule_map.begin();
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: again, please drop this line.

File: src/rgw/rgw_json_enc.cc [hunk 16]
```diff
@@ -1494,3 +1494,76 @@ void RGWOrphanSearchState::dump(Formatter *f) const
...
+  f->open_array_section("rule_map");
+  for (auto& multi_iter : rule_map) {
```
Comment (tchaikov) [inline src/rgw/rgw_json_enc.cc]: s/multi_iter/rule/

**Comment by cooboos** [comment]
@tchaikov  check pls.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-11-28_10:23:40-rados-wip-kefu-testing-2017-11-28-1542-distro-basic-mira/

failures tracked/fixed by

- http://tracker.ceph.com/issues/22093
- http://tracker.ceph.com/issues/22043
- #19225

**Pull request merged**